// igniteui-angular/esm2015/lib/services/transaction/transaction-factory.service.js
import { Injectable } from '@angular/core';
import { IgxBaseTransactionService } from './base-transaction';
import { IgxHierarchicalTransactionService } from './igx-hierarchical-transaction';
import { IgxTransactionService } from './igx-transaction';
import * as i0 from "@angular/core";
/**
 * Factory service for instantiating TransactionServices
 */
export class IgxFlatTransactionFactory {
    /**
     * Creates a new Transaction service instance depending on the specified type.
     *
     * @param type The type of the transaction
     * @returns a new instance of TransactionService<Transaction, State>
     */
    create(type) {
        switch (type) {
            case ("Base" /* Base */):
                return new IgxTransactionService();
                ;
            default:
                return new IgxBaseTransactionService();
        }
    }
}
IgxFlatTransactionFactory.ɵprov = i0.ɵɵdefineInjectable({ factory: function IgxFlatTransactionFactory_Factory() { return new IgxFlatTransactionFactory(); }, token: IgxFlatTransactionFactory, providedIn: "root" });
IgxFlatTransactionFactory.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/**
 * Factory service for instantiating HierarchicalTransactionServices
 */
export class IgxHierarchicalTransactionFactory extends IgxFlatTransactionFactory {
    /**
     * Creates a new HierarchialTransaction service instance depending on the specified type.
     *
     * @param type The type of the transaction
     * @returns a new instance of HierarchialTransaction<HierarchialTransaction, HierarchialState>
     */
    create(type) {
        switch (type) {
            case ("Base" /* Base */):
                return new IgxHierarchicalTransactionService();
                ;
            default:
                return new IgxBaseTransactionService();
        }
    }
}
IgxHierarchicalTransactionFactory.ɵprov = i0.ɵɵdefineInjectable({ factory: function IgxHierarchicalTransactionFactory_Factory() { return new IgxHierarchicalTransactionFactory(); }, token: IgxHierarchicalTransactionFactory, providedIn: "root" });
IgxHierarchicalTransactionFactory.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
